openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,428 @@
+    return inopt
+
+
+cdef size_t write_ogg_page(size_t offset, bytearray outdata, ogg.ogg_page *og):
+    '''
+    Write og at offset into outdata.
+    If succesful, returns new offset, else 0.
+    '''
+    if PyByteArray_GET_SIZE(outdata) < offset + og.header_len + og.body_len:
+        try:
+            PyByteArray_Resize(outdata, offset + og.header_len + og.body_len)
+            # TODO increase more than necessary
+        except MemoryError:
+            return 0
+
+    cdef char *out = PyByteArray_AS_STRING(outdata) + offset
+    memcpy(out, og.header, og.header_len)
+    memcpy(out + og.header_len, og.body, og.body_len)
+    return offset + og.header_len + og.body_len
+
+
+cdef void free_buffers(ogg.ogg_stream_state *os, opus.OpusEncoder *oe,
+                       unsigned char *pymem1, unsigned char *pymem2):
+    ogg.ogg_stream_clear(os)
+    if oe != NULL:
+        opus.opus_encoder_destroy(oe)
+    if pymem1 != NULL:
+        PyMem_Free(pymem1)
+    if pymem2 != NULL:
+        PyMem_Free(pymem2)
+    return
+
+
+cdef upsample(indata, dict opts, int target_rate):
+    '''
+    Upsamples the PCM-data in 'indata' to 'target_rate' using linear
+    interpolation. 'opts' is a dict containing info on 'indata'.
+    If successful, return the upsampled data, otherwise an error message.
+    Assumptions: * 1 ≤ opts['bit_depth'] ≤ 16
+                 * opts['input_rate'] < target_rate
+                 * opts['channels'] ∈ {1, 2}
+    TODO: * a more eloquent version?
+          * a faster version?
+          * a better interpolation?
+    '''
+    cdef:
+        int wframe_len = opts['wframe_len']
+        int alpha, ch, channel = opts['channels']
+        int bit_depth = opts['bits_per_sample']
+        unsigned short mask = ~((1 << (16 - bit_depth)) - 1)
+
+        size_t iin = 0, iout = 0
+        size_t ismpls, osmpls  # number of input and output samples
+        short a,b
+        int num, den, divisor
+
+        size_t outsize
+        const char *inp = indata
+        char *out
+
+    ismpls = len(indata) // wframe_len
+    if ismpls < 2:
+        # Nothing to do.
+        return indata
+
+    divisor = gcd(target_rate, opts['input_rate'])
+    num = <int> (target_rate / divisor)
+    den = <int> (opts['input_rate'] / divisor)
+
+    # We're using linear interpolation. That means, we need for every output
+    # sample one input samples that comes before and one after the output
+    # sample. ismpls - 1 is the index of the last input sample. By multiplying
+    # with num / den we get the index of the last output sample. In general,
+    # this is not a whole number. Therefore we take only the integer part.
+    # Because indices are zero-based, add one for the number of output samples.
+    osmpls = ((ismpls - 1) * num / den).__trunc__() + 1
+    outsize = osmpls * channel * 2  # we convert to int16
+
+    try:
+        ret = bytearray(outsize)
+    except MemoryError:
+        return "Could not allocate buffer."
```

Comment:
I think you can safely just let that exception raise, other memory errors are raised anyway.